[FEATURE] Retirer l'accès à la page "Profils cibles" de Pix Admin uniquement au rôle "CERTIF" (PIX-4668)

igorissen opened the pull request:
## :unicorn: Problème

La page `Profils cibles` dans l'application Pix Admin est visible à tous les utilisateurs de l'application indépendamment de leur rôles. Seul des utilisateurs ayant le rôle `CERTIF` ne devrait pas avoir accès à cette page.

![Screenshot 2022-05-30 at 09 18 44](https://user-images.githubusercontent.com/6919604/170938288-924f767d-536f-4e16-855a-bcd85e9308dd.png)

## :robot: Solution

1. Bloquer l'accès à toutes les routes liées à l'onglet `Profils cibles` au niveau de l'API.
2. Ne pas afficher l'onglet `Profils cibles` dans la barre des menus.
3. Rediriger l'utilisateur vers la page `Organisations` si ce dernier tente d'utiliser directement l'URL vers les pages associées à l'onglet `Profils cibles`.

## :rainbow: Remarques

Des routes ont été déplacé sous la route `/admin` :

- `/api/frameworks` >> `/api/admin/frameworks`
- `/api/frameworks/{id}/areas` >> `/api/admin/frameworks/{id}/areas`

## :100: Pour tester

#### Avec un utilisateur ayant le rôle `SUPER_ADMIN`, `SUPPORT` ou `METIER`

 - Se connecter avec `superadmin@example.net / pix123`, `pixsupport@example.net / pix123` ou `pixmetier@example.net / pix123`
- Voir au niveau du menu à gauche que l'onglet `Profils cibles` est présent
- Remplacer le chemin `/organizations/list` dans la barre de navigation du navigateur par `/target-profiles/list`
- Constater que l'on arrive sur la page `Profils cibles`
- Utiliser l'application
- Constater que toutes les actions de l'onglet `Profils cibles` fonctionnent.

#### Avec un utilisateur ayant le rôle `CERTIF`

- Se connecter avec `pixcertif@example.net / pix123`
- Voir au niveau du menu à gauche que l'onglet `Profils cibles` n'est pas présent
- Remplacer le chemin `/organizations/list` dans la barre de navigation du navigateur par `/target-profiles/list`
- Constater que l'on arrive sur la page `Organisations`

igorissen committed:
feat(api): restrict access to get badge resource

- GET /api/admin/badges/{id}
- A user with a role "CERTIF" does not have access to this resource

igorissen committed:
feat(api): restrict access to delete badge resource

- DELETE /api/admin/badges/{id}
- A user with a role "CERTIF" does not have access to this resource

igorissen committed:
feat(api): restrict access to create stage resource

- POST /api/admin/stages
- A user with a role "CERTIF" does not have access to this resource

igorissen committed:
feat(api): restrict access to update stage resource

- PATCH /api/admin/stages/{id}
- A user with a role "CERTIF" does not have access to this resource

igorissen committed:
refactor(api): reorder routes

igorissen committed:
feat(api): restrict access to get target profiles resource

- GET /api/admin/target-profiles
- A user with a role "CERTIF" does not have access to this resource

igorissen committed:
feat(api): restrict access to create target profiles resource

- POST /api/admin/target-profiles
- A user with a role "CERTIF" does not have access to this resource

igorissen committed:
feat(api): restrict access to get target profile resource

- GET /api/admin/target-profiles/{id}
- A user with a role "CERTIF" does not have access to this resource

igorissen committed:
feat(api): restrict access to update target profile resource

- PATCH /api/admin/target-profiles/{id}
- A user with a role "CERTIF" does not have access to this resource

igorissen committed:
feat(api): restrict access to outdate target profile resource

- PUT /api/admin/target-profiles/{id}/outdate
- A user with a role "CERTIF" does not have access to this resource

pix-service commented:
I'm deploying this PR to these urls:

- App (.fr): https://app-pr4455.review.pix.fr
- App (.org): https://app-pr4455.review.pix.org
- Orga: https://orga-pr4455.review.pix.fr
- Certif: https://certif-pr4455.review.pix.fr
- Admin: https://admin-pr4455.review.pix.fr
- API: https://api-pr4455.review.pix.fr/api/

 Please check it out!

igorissen committed:
feat(api): restrict access to simply access of target profile resource

- PUT /api/admin/target-profiles/{id}/simplified-access
- A user with a role "CERTIF" does not have access to this resource

igorissen committed:
feat(api): restrict access to get target profile organizations resource

- GET /api/admin/target-profiles/{id}/organizations
- A user with a role "CERTIF" does not have access to this resource

igorissen committed:
feat(api): restrict access to get target profile stages resource

- GET /api/admin/target-profiles/{id}/stages
- A user with a role "CERTIF" does not have access to this resource

igorissen committed:
feat(api): restrict access to get target profile badges resource

- GET /api/admin/target-profiles/{id}/badges
- A user with a role "CERTIF" does not have access to this resource

igorissen committed:
feat(api): restrict access to create target profile badge resource

- POST /api/admin/target-profiles/{id}/badges
- A user with a role "CERTIF" does not have access to this resource

igorissen committed:
feat(api): restrict access to attach an organization to target profile resource

- POST /api/admin/target-profiles/{id}/attach-organizations
- A user with a role "CERTIF" does not have access to this resource

igorissen committed:
feat(api): restrict access to copy an organization to target profile resource

- POST /api/admin/target-profiles/{id}/copy-organizations
- A user with a role "CERTIF" does not have access to this resource

igorissen committed:
feat(admin): restrict access to "Profils cibles" page

- A user with a role "CERTIF" does not have access to this page

igorissen committed:
feat(admin): restrict access to target profiles list page

igorissen committed:
feat(admin): restrict access to tubes selection page

igorissen committed:
feat(admin): restrict access to target profile creation page

igorissen committed:
feat(admin): restrict access to stage details page

igorissen committed:
feat(admin): restrict access to badge details page

igorissen committed:
feat(admin): restrict access to target profile details sub-pages

igorissen committed:
feat(admin): restrict access to target profile details page

igorissen committed:
refactor(api): move /frameworks routes under /admin

igorissen committed:
feat(admin): add framework adapter to use /api/admin namespace

annemarie35 commented on admin/app/services/access-control.js:
```diff
@@ -9,6 +9,14 @@ export default class AccessControlService extends Service {
     return this.currentUser.adminMember.isSuperAdmin || this.currentUser.adminMember.isSupport;
   }
 
+  get hasAccessToTargetProfilesActionsScope() {
```
on pourrait tester unitaire ce service et faire un seul test d'intégration dans `admin/tests/integration/components/menu-bar_test.js` plutôt que de retester les 4 rôles

annemarie35 commented on admin/tests/unit/routes/authenticated/target-profiles/list_test.js:
```diff
@@ -113,4 +114,73 @@ module('Unit | Route | authenticated/target-profiles/list', function (hooks) {
+      router.transitionTo = sinon.stub();
+    });
+
+    test('it should allow an admin member with "SUPER_ADMIN" role to access the page', function (assert) {
```
Je me pose des questions sur cette façon de tester car on est en train de tester finalement le comportement de `this.accessControl.restrictAccessTo` et sans le stubber.
Il faudrait plutôt faire comme ça `admin/tests/unit/routes/authenticated/organizations/new_test.js` et vérifier que dans le beforeModel on appelle bien `restrictAccessTo` avec les bons paramètres

annemarie35 commented on api/lib/application/badges/index.js:
```diff
@@ -27,7 +26,7 @@ exports.register = async function (server) {
+        tags: ['api', 'admin', 'badges'],
         notes: [
           "- **Cette route est restreinte aux utilisateurs authentifiés ayant les droits d'accès**\n" +
             '- Elle permet de récupérer un résultat thématique.',
```
Un /badge c'est donc un `résultat thématique` ^^ ?

igorissen replied:
Il semblerait que oui

annemarie35 commented on api/lib/application/badges/index.js:
```diff
@@ -27,7 +26,7 @@ exports.register = async function (server) {
+        tags: ['api', 'admin', 'badges'],
         notes: [
           "- **Cette route est restreinte aux utilisateurs authentifiés ayant les droits d'accès**\n" +
             '- Elle permet de récupérer un résultat thématique.',
```
Un /badge c'est donc un `résultat thématique` ^^ ?

annemarie35 commented on api/tests/unit/application/badges/index_test.js:
```diff
@@ -0,0 +1,76 @@
+      await httpTestServer.register(badgesRouter);
+
+      // when
+      const { statusCode } = await httpTestServer.request('GET', '/api/admin/badges/invalid');
```
j'ai cru qu'on avait une route `/api/admin/badges/invalid` ^^ je suggère de mettre `/api/admin/badges/invalidId`

annemarie35 replied:
j'ai cru qu'on avait une route `/api/admin/badges/invalid` ^^ je suggère de mettre `/api/admin/badges/invalidId`

annemarie35 commented on admin/app/routes/authenticated/target-profiles/tubes-selection.js:
```diff
@@ -3,6 +3,11 @@ import { inject as service } from '@ember/service';
 
 export default class TubesSelectionRoute extends Route {
```
@igorissen il manque côté api la route http://localhost:4202/target-profiles/tubes-selection 
J'ai pas trouvé fonctionnellement où c'était appelé

annemarie35 replied:
Remarque, ça serait plus visible à mes yeux si on mettait un dossier /admin dans `api/lib/application` au lieu de chercher via le repo admin les routes qui sont liées à admin dans api/lib/application

annemarie35 replied:
Remarque, ça serait plus visible à mes yeux si on mettait un dossier /admin dans `api/lib/application` au lieu de chercher via le repo admin les routes qui sont liées à admin dans api/lib/application

annemarie35 replied:
@Anne-Gaelle-S @iwazaru  le sujet a déjà été evoqué chez Pix ?

annemarie35 replied:
@Anne-Gaelle-S @iwazaru  le sujet a déjà été evoqué chez Pix ?

annemarie35 commented on admin/app/routes/authenticated/target-profiles/tubes-selection.js:
```diff
@@ -3,6 +3,11 @@ import { inject as service } from '@ember/service';
 
 export default class TubesSelectionRoute extends Route {
```
@igorissen il manque côté api la route http://localhost:4202/target-profiles/tubes-selection 
J'ai pas trouvé fonctionnellement où c'était appelé

annemarie35 replied:
y'a une route `path: '/api/tubes/{id}/skills',` qui est accessible seulement aux super admin, ça vaudrait peut-être le coup de vérifier où elle est utilisée car c'est lié au badges et target_profiles

igorissen replied:
Oui, je viens de voir que je l'avais oublié. C'est la génération du JSON sur la page tubes-selection.

igorissen replied:
La page `tubes-selection` fait appel à la route `/api/admin/frameworks`

annemarie35 commented on api/tests/unit/application/target-profiles/index_test.js:
```diff
@@ -5,7 +5,64 @@ const targetProfileController = require('../../../../lib/application/target-prof
+
+    beforeEach(function () {
+      method = 'GET';
+      url = '/api/admin/target-profiles';
```
On teste dans ce describe le GET `/api/admin/target-profiles`, je pense qu'on pourrait se passer du before each et mettre method et url en cons non ? Qu'en penses-tu ?

igorissen replied:
Yes, on peut faire ça.

annemarie35 commented on api/tests/unit/application/target-profiles/index_test.js:
```diff
@@ -64,11 +115,8 @@ describe('Integration | Application | Target Profiles | Routes', function () {
+      const response = await httpTestServer.request(method, `${url}?page[size]=azerty`);
 
       // then
       expect(response.statusCode).to.equal(400);
```
Pourquoi ne pas remplacer tous les `const response =`  par `{statusCode} = ` 
C'est pas bloquant mais on se retrouve avec deux styles maintenant ^^

annemarie35 commented on api/tests/unit/application/target-profiles/index_test.js:
```diff
@@ -160,15 +165,61 @@ describe('Integration | Application | Target Profiles | Routes', function () {
-
       // when
-      const response = await httpTestServer.request(method, url);
+      const response = await httpTestServer.request(method, '/api/admin/target-profiles/azerty');
```
On a utilisé let url et là c'était l'occasion de lui réaffecter une valeur et justifier le beforeEach.
Je propose soit de remettre comme c'était, soit d'enlever aussi le beforeEach et de définir les valeurs d'url et method via des const

Autre remarque, le test qui s'appelle 'should resolve with correct id' ligne 147  peut être supprimé car ça ne teste pas plus que 'should return a response with an HTTP status code 200 when user has role "SUPER_ADMIN", "SUPPORT", "METIER"' (voir ça teste moins car on a pas utilisé le withArgs dans le stub)

annemarie35 commented on api/tests/unit/application/target-profiles/index_test.js:
```diff
@@ -697,16 +697,15 @@ describe('Integration | Application | Target Profiles | Routes', function () {
 
-      const method = 'PATCH';
-      const payload = {
+    beforeEach(function () {
```
Dans l'équipe, on essaye plutôt de se débarrasser des beforeEach quand ils ne sont pas nécessaires ou qu'ils apportent de la complexité dans la lecture

annemarie35 commented on api/tests/unit/application/target-profiles/index_test.js:
```diff
@@ -780,51 +778,64 @@ describe('Integration | Application | Target Profiles | Routes', function () {
-        },
-      };
-      const url = '/api/admin/target-profiles/9999999';
+    it('should return a response with an HTTP status code 204 when user has role "SUPER_ADMIN", "SUPPORT" or "METIER"', async function () {
```
le test `it('should exist',` en ligne 986 peut être supprimé

annemarie35 commented on api/tests/unit/application/target-profiles/index_test.js:
```diff
@@ -839,44 +839,76 @@ describe('Integration | Application | Target Profiles | Routes', function () {
+      payload = {};
+    });
+
     it('should exist', async function () {
```
ce test `it('should exist'` peut être suprimé

annemarie35 commented on api/tests/unit/application/target-profiles/index_test.js:
```diff
@@ -817,44 +817,70 @@ describe('Integration | Application | Target Profiles | Routes', function () {
-      const method = 'PUT';
-      const payload = {};
-      const url = '/api/admin/target-profiles/9999999/outdate';
+    it('should return a response with an HTTP status code 204 when user has role "SUPER_ADMIN", "SUPPORT" or "METIER"', async function () {
```
il manque un test sur l'erreur 400 produite quand ça ne passe pas la validation joi
```
  params: Joi.object({
            id: identifiersType.targetProfileId,
          }),
```

annemarie35 commented on api/tests/unit/application/target-profiles/index_test.js:
```diff
@@ -883,4 +883,75 @@ describe('Integration | Application | Target Profiles | Routes', function () {
     });
   });
+
+  describe('PUT /api/admin/target-profiles/{id}/simplified-access', function () {
```
Même remarques que précédemment 
- il manque un test sur la 400 de joi
- le test it should exist n'apporte rien de plus et peut être supprimé

annemarie35 commented on api/tests/unit/application/target-profiles/index_test.js:
```diff
@@ -212,7 +212,10 @@ describe('Integration | Application | Target Profiles | Routes', function () {
     });
   });
```
On peut supprimer le beforeEach et mettre des const ici car url ne change pas de valeur

igorissen replied:
Je l'ai fait sur les commits suivants.

igorissen committed:
feat(api): restrict access to get tube skills resource

- GET /api/admin/tubes/{id}/skills
- A user with a role "CERTIF" does not have access to this resource

annemarie35 commented on api/tests/unit/application/target-profiles/index_test.js:
```diff
@@ -403,6 +403,72 @@ describe('Integration | Application | Target Profiles | Routes', function () {
+      await httpTestServer.register(moduleUnderTest);
+
+      // when
+      const { statusCode } = await httpTestServer.request('GET', '/api/admin/target-profiles/azerty/badges');
```
chippotage : mettre method à la place de GET

annemarie35 commented on api/tests/unit/application/target-profiles/index_test.js:
```diff
@@ -766,89 +766,98 @@ describe('Integration | Application | Target Profiles | Routes', function () {
+      },
+    };
+
     it('should return 201 HTTP response', async function () {
```
ce test `it('should return 201 HTTP response'` peut être supprimé car celui qui teste avec les roles fait la même chose en plus complet

annemarie35 commented on api/tests/unit/application/target-profiles/index_test.js:
```diff
@@ -606,90 +606,110 @@ describe('Integration | Application | Target Profiles | Routes', function () {
-
       // when
-      const response = await httpTestServer.request(method, url, payload);
+      const { statusCode } = await httpTestServer.request(method, url, payload);
```
le test `it('should resolve with correct id'` peut être supprimé

annemarie35 commented on api/tests/unit/application/target-profiles/index_test.js:
```diff
@@ -723,66 +727,89 @@ describe('Integration | Application | Target Profiles | Routes', function () {
-
       // when
-      const response = await httpTestServer.request(method, url, payload);
+      const { statusCode } = await httpTestServer.request(method, url, payload);
```
le test `it('should resolve', async function () {` peut être supprimé

annemarie35 commented on admin/tests/integration/components/menu-bar_test.js:
```diff
@@ -7,13 +7,43 @@ module('Integration | Component | menu-bar', function (hooks) {
   setupRenderingTest(hooks);
 
   test('should contain link to "organizations" management page', async function (assert) {
```
En lisant ces tests, je n'arrive pas à savoir si `link to "organizations" management page` est réservé seulement aux super admin ou si on a fait un seul test. 
Je trouverais ça plus lisible de mettre du contexte comme sur `module('Team tab', function () {`
C'est un peu hors scope de cette PR mais si tu as la connaissance en test ça serait cool je pense pour la suite pour savoir précisément les droits pour chaque élément du menu et que ça soit clair quand c'est juste les super admin qui voient un élément du menu. ^^

igorissen replied:
Oui, on pourrait faire cette modification dans un ticket tech. C'est vrai que pour les parties où nous n'avons pas fait de modification ce n'est pas clair.

annemarie35 reviewed (commented):
Je ne vois aucun test d'acceptance pour tester la redirection sur toutes les routes protégées, c'est voulu ?

Question, dans le router, on voit qu'il y a une route insight

```
    this.route('target-profiles', function () {
      this.route('list');
      this.route('new');
      this.route('tubes-selection');
      this.route('target-profile', { path: '/:target_profile_id' }, function () {
        this.route('details', { path: '/' });
        this.route('organizations');
        this.route('insights');
        this.route('badges', function () {
          this.route('new');
        });
      });
    });
```
Elle n'a pas été protégée mais on n'y accède pas (on est redirigée), j'arrive pas à voir par quel mécanisme ember, si quelqu'un(e) a la réponse, je veux bien ^^

AndreiaPena commented:
Testé fonctionnellement : 
Connecté avec un rôle CERTIF:
-  je ne vois pas le profil cible dans le menu ✅ 
- je force l'url avec `target-profiles/list`/ `target-profiles/1` / `target-profiles/1/organizations`, tout me renvoi vers la liste des organisations ✅ 

Connecté avec un rôle ADMIN/METIER/SUPPORT : 
- j'ai accès au profil cible du menu ✅ 
- Les actions et accès aux différentes pages des profils cibles passent tous ✅

AndreiaPena commented on api/tests/unit/application/badges/index_test.js:
```diff
@@ -0,0 +1,76 @@
+
+describe('Unit | Application | Badges | Routes', function () {
+  describe('GET /api/admin/badges/{id}', function () {
+    it('should return a response with an HTTP status code 200 when user has role "SUPER_ADMIN", "SUPPORT" or "METIER"', async function () {
```
**chipotage/suggestion** : Si on se retrouve à expliciter un when dans le nom du test, alors autant créer un describe plus haut pour porter ce "when", la lecture sera plus simple.

exemple : 
```
describe('when user has role "SUPER_ADMIN", "SUPPORT" or "METIER"', function () {
    it('should return a response with an HTTP status code 200', async function () {
```

AndreiaPena replied:
Tous les tests de ce ticket sont concernés

AndreiaPena commented on api/tests/unit/application/stages/index_test.js:
```diff
@@ -3,24 +3,53 @@ const securityPreHandlers = require('../../../../lib/application/security-pre-ha
+          securityPreHandlers.checkUserHasRoleSupport,
+          securityPreHandlers.checkUserHasRoleMetier,
+        ])
+        .callsFake(() => (request, h) => h.response(true));
```
**Question/Remarque** : Utiliser `withArgs` dans le test pour vérifier que ce sont les bons arguments ✅  mais je me demande si c'est pertinent de le faire à tous les tests suivants. Si on le vérifie une fois dans un test, ça suffirait non ?

Puis les autres n'auraient que : 
`sinon.stub(securityPreHandlers, 'userHasAtLeastOneAccessOf').returns(() => true);`
(pour ceux qui retournent true évidemment, sinon callsFake pour les autres)

Gain de ligne de code et facilité de lecture des tests.

AndreiaPena commented on api/lib/application/target-profiles/index.js:
```diff
@@ -92,24 +92,21 @@ exports.register = async (server) => {
-        handler: targetProfileController.attachOrganizations,
-        tags: ['api', 'target-profiles'],
+        handler: targetProfileController.findTargetProfileBadges,
+        tags: ['api', 'target-profiles', 'badges'],
```
**Chipotage** : "admin" en plus ? (comme tu l'as fait dans les autres routes, je m'y suis habituée haha)

igorissen replied:
Normalement ce changement devrait se trouver dans les commits suivants. Mais pour être sûr, je vais faire une repasse sur les commits.

AndreiaPena commented on api/lib/application/target-profiles/index.js:
```diff
@@ -124,21 +121,56 @@ exports.register = async (server) => {
         notes: [
           "- **Cette route est restreinte aux utilisateurs authentifiés ayant les droits d'accès**\n" +
-            '- Elle permet de rattacher à un profil cible donné les organisations d’un profil cible existant (id de ce dernier en payload)',
+            '- Elle permet de récupérer les organizations auxquelles est rattaché le profil cible',
```
**Chipotage/hors PR** : "les organisations"

AndreiaPena commented on api/lib/application/target-profiles/index.js:
```diff
@@ -209,7 +208,7 @@ exports.register = async (server) => {
         },
         handler: targetProfileController.createTargetProfile,
-        tags: ['api', 'target-profiles', 'create'],
+        tags: ['api', 'admin ', 'target-profiles', 'create'],
```
**Chipotage** : 'admin' sans espace

AndreiaPena commented on api/tests/unit/application/target-profiles/index_test.js:
```diff
@@ -802,44 +802,68 @@ describe('Unit | Application | Target Profiles | Routes', function () {
-      const method = 'PUT';
-      const payload = {};
-      const url = '/api/admin/target-profiles/9999999/outdate';
+    it('should reject request with HTTP code 400', async function () {
```
**Remarque** : Il manque la raison de ce 400 dans le nom ( privilégier les when dans un describe que dans le test même)

AndreiaPena commented on api/tests/unit/application/target-profiles/index_test.js:
```diff
@@ -866,4 +866,74 @@ describe('Unit | Application | Target Profiles | Routes', function () {
+      expect(statusCode).to.equal(200);
+    });
+
+    it('should reject request with HTTP code 400', async function () {
```
**Remarque** : Pareil ici

AndreiaPena commented on api/tests/unit/application/target-profiles/index_test.js:
```diff
@@ -240,60 +240,99 @@ describe('Unit | Application | Target Profiles | Routes', function () {
-
       // when
-      const response = await httpTestServer.request(method, url);
+      const { statusCode } = await httpTestServer.request('GET', '/api/admin/target-profiles/azerty/organizations');
```
**Chipotage/Remarque** : C'est un choix d'écriture, les deux me conviennent mais ici on à la méthode GET en dur et les autres bénéficient de la constante plus haut.

igorissen replied:
Un oublie de ma part.

AndreiaPena commented on admin/tests/integration/components/menu-bar_test.js:
```diff
@@ -7,13 +7,43 @@ module('Integration | Component | menu-bar', function (hooks) {
   });
 
+  module('Target Profiles tab', function () {
+    test('should contain link to "Target Profiles" page when admin member have "SUPER_ADMIN" as role', async function (assert) {
```
**Remarque** : ici je comprend que seul le super admin à accès mais c'est le cas de métier et support aussi. Même si ici on teste superadmin, on peut au moins le spécifier dans le titre :)

igorissen replied:
Oui, tout à fait raison.

AndreiaPena commented on admin/tests/unit/services/access-control-test.js:
```diff
@@ -77,6 +77,48 @@ module('Unit | Service | access-control', function (hooks) {
+    test('should be false if admin member has role "CERTIF"', function (assert) {
+      // given
+      const currentUser = this.owner.lookup('service:current-user');
+      currentUser.adminMember = { isSuperAdmin: false, isSupport: false, isMetier: false };
```
**Question** : Pourquoi passer par ça plutôt que `isCertif: true` ?

AndreiaPena commented on api/lib/application/frameworks/index.js:
```diff
@@ -33,22 +29,48 @@ exports.register = async function (server) {
+  ];
+
+  server.route([
+    ...adminRoutes,
```
**Question** : C'est un choix d'équipe de séparer les routes admin de cette façon ?

igorissen replied:
Non ce n'est pas un choix d'équipe. L'envie est plus d'avoir un dossier `/admin` et y mettre toutes les routes concernant admin mais ça risque de faire un gros changement. Pour faire cette distinction plus facilement, et Yasmina ayant fait la même modification, je me suis permis de faire de même.

Je trouve que c'est déjà plus claire la séparation entre les routes admin et les autres. Cette distinction devra être faite aussi pour les autres applications si certaines routes sont uniques à chacune des applications.

AndreiaPena commented on api/lib/application/frameworks/index.js:
```diff
@@ -33,22 +29,48 @@ exports.register = async function (server) {
-        tags: ['api', 'framework'],
+        tags: ['api', 'admin', 'framework'],
         notes: [
           'Cette route est restreinte aux utilisateurs authentifiés avec le rôle Super Admin',
```
**Remarque** : Cette note n'est du coup plus bonne. "Cette route est restreinte aux utilisateurs authentifiés avec le rôle Super Admin, Support ou Métier"

AndreiaPena commented on api/lib/application/frameworks/index.js:
```diff
@@ -4,27 +4,23 @@ const securityPreHandlers = require('../security-pre-handlers');
+        ],
+        tags: ['api', 'admin', 'framework'],
         notes: [
           'Cette route est restreinte aux utilisateurs authentifiés avec le rôle Super Admin',
```
**Remarque** : Il faut changer cette phrase au profit de "Cette route est restreinte aux utilisateurs authentifiés avec le rôle Super Admin, Support ou Métier"

AndreiaPena commented on api/lib/application/frameworks/index.js:
```diff
@@ -33,22 +29,48 @@ exports.register = async function (server) {
     {
       method: 'GET',
-      path: '/api/frameworks/{id}/areas',
+      path: '/api/admin/frameworks/{id}/areas',
```
**Remarque/Question** : J'ai souvenir qu'on voulait changer une route déjà existante mais quelqu'un avait suggérer de plutôt la dupliquer plutôt que de changer direct. Je ne sais pas si le "risque" se pose ici. @Anne-Gaelle-S ça te parle ?
PR concerné https://github.com/1024pix/pix/pull/4282

igorissen replied:
J'ai pris contact avec l'équipe de team contenu pour vérifier que cette route n'était utilisée que pour l'admin avant de faire la modification.

Anne-Gaelle-S replied:
oui c'était Vincent je crois qui suggérait ça, pour éviter lors de la mise en prod qu'il y ait un décalage entre le front et l'api je crois 🤔 (après ici c'est Pix Admin, donc l'utilisation est plus restreinte)

AndreiaPena commented on api/tests/unit/application/frameworks/index_test.js:
```diff
@@ -0,0 +1,111 @@
+const frameworksController = require('../../../../lib/application/frameworks/frameworks-controller');
+const moduleUnderTest = require('../../../../lib/application/frameworks');
+
+describe('Integration | Application | Frameworks | Routes', function () {
```
Unit

AndreiaPena commented on api/lib/application/tubes/index.js:
```diff
@@ -7,18 +7,27 @@ exports.register = async function (server) {
+        tags: ['api', 'admin', 'tubes', 'skills'],
         notes: [
-          'Cette route est restreinte aux utilisateurs authentifiés avec le rôle Super Admin',
+          'Cette route est restreinte aux utilisateurs authentifiés avec le rôle Super Admin, Support ou Metier',
```
**Remarque** : Il faudrait qu'on soit iso sur cette phrase là pour toutes les routes concernées.
+ coquille => Métier

igorissen commented:
> Je ne vois aucun test d'acceptance pour tester la redirection sur toutes les routes protégées, c'est voulu ?
> 
> Question, dans le router, on voit qu'il y a une route insight
> 
> ```
>     this.route('target-profiles', function () {
>       this.route('list');
>       this.route('new');
>       this.route('tubes-selection');
>       this.route('target-profile', { path: '/:target_profile_id' }, function () {
>         this.route('details', { path: '/' });
>         this.route('organizations');
>         this.route('insights');
>         this.route('badges', function () {
>           this.route('new');
>         });
>       });
>     });
> ```
> 
> Elle n'a pas été protégée mais on n'y accède pas (on est redirigée), j'arrive pas à voir par quel mécanisme ember, si quelqu'un(e) a la réponse, je veux bien ^^

@annemarie35 J'ai mis la protection sur le fichier `/authenticated/target-profiles/target-profile/index` qui est le parent des sous pages `details`, `insights` et `organizations`.

Pour les tests d'acceptance manquants, c'est corrigé.
